### ISSUE TITLE
Posting to mqtt gets OSErrors

### ISSUE THREAD

**Issue by lubarb** [adafruit/Adafruit_CircuitPython_MiniMQTT#77]
I'm running on a Magtag using "Adafruit CircuitPython 6.2.0-beta.4 on 2021-03-18; Adafruit MagTag with ESP32S2" and libraries from 3/20/2021, except for a modified adafruit_tlv493d which has a pull request in.  I'm reading data from the tlv and posting the data to a dedicated raspberry pi broker in a loop without delays. Within a handful of seconds I get OSErrors such as:

```
Traceback (most recent call last):
  File "code.py", line 99, in <module>
  File "adafruit_minimqtt/adafruit_minimqtt.py", line 616, in publish
OSError: [Errno 11] EAGAIN
```
and
```
Traceback (most recent call last):
  File "code.py", line 99, in <module>
  File "adafruit_minimqtt/adafruit_minimqtt.py", line 614, in publish
OSError: [Errno 104] ECONNRESET
```
I had gotten an error like this posting values once a second overnight, which is why I tried running with no delay.

Also MQTT Explorer on Windows sometimes reports scrambled results after the crash, such as:

`{"z": 0.0, "y":06dryertest/doo`

```
import board
import busio
import wifi
import socketpool
import ssl
import time
import json
import binascii
import struct
import adafruit_minimqtt.adafruit_minimqtt as MQTT
import adafruit_tlv493d

DOOR_MQTT_TOPIC = "dryertest/door"

print("starting up")

# Hardware I2C setup:
#i2c = busio.I2C(board.SCL, board.SDA, frequency = 400000)
i2c = busio.I2C(board.SCL, board.SDA)
tlv = adafruit_tlv493d.TLV493D(i2c)

# Add a secrets.py to your filesystem that has a dictionary called secrets with "ssid" and
# "password" keys with your WiFi credentials. DO NOT share that file or commit it into Git or other
# source control.
# pylint: disable=no-name-in-module,wrong-import-order
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets are kept in secrets.py, please add them there!")
    raise

wifi.radio.connect(secrets["ssid"], secrets["password"])
print("Connected to %s!" % secrets["ssid"])

# Create a socket pool
pool = socketpool.SocketPool(wifi.radio)

# Set up a MiniMQTT Client
mqtt_client = MQTT.MQTT(
    broker=secrets["mqtt_broker"],
    port=secrets["mqtt_port"],
    username=secrets["mqtt_username"],
    password=secrets["mqtt_password"],
    socket_pool=pool,
    ssl_context=ssl.create_default_context(),
)

print("Attempting to connect to %s" % mqtt_client.broker)
mqtt_client.connect()

readcount = 0

while True:
    door = tlv.magnetic

    door_output = {
    "x" : door[0],
    "y" : door[1],
    "z" : door[2],
    }

    readcount += 1
    if (readcount % 100 == 1) :
        print('.', end='')
    if (readcount % 1000 == 1) :
        print("\n read: : ", readcount)
    mqtt_client.publish(DOOR_MQTT_TOPIC, json.dumps(door_output))

#    time.sleep(1)
```

**Comment by lubarb** [comment]
I simplified the code a bit and have discovered that I get the errors even if I'm not accessing the tlv inside the loop. I've gotten errors at < 2000 'reads', but sometimes it also takes 8000 or so....

```
import board
import busio
import wifi
import socketpool
import ssl
import time
import json
import binascii
import struct
import adafruit_minimqtt.adafruit_minimqtt as MQTT
import adafruit_tlv493d

DOOR_MQTT_TOPIC = "dryertest/door"

print("starting up")

# Hardware I2C setup:
i2c = busio.I2C(board.SCL, board.SDA)
tlv = adafruit_tlv493d.TLV493D(i2c)

# Add a secrets.py to your filesystem that has a dictionary called secrets with "ssid" and
# "password" keys with your WiFi credentials. DO NOT share that file or commit it into Git or other
# source control.
# pylint: disable=no-name-in-module,wrong-import-order
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets are kept in secrets.py, please add them there!")
    raise

wifi.radio.connect(secrets["ssid"], secrets["password"])
print("Connected to %s!" % secrets["ssid"])

# Create a socket pool
pool = socketpool.SocketPool(wifi.radio)

# Set up a MiniMQTT Client
mqtt_client = MQTT.MQTT(
    broker=secrets["mqtt_broker"],
    port=secrets["mqtt_port"],
    username=secrets["mqtt_username"],
    password=secrets["mqtt_password"],
    socket_pool=pool,
    ssl_context=ssl.create_default_context(),
)

print("Attempting to connect to %s" % mqtt_client.broker)
mqtt_client.connect()

readcount = 0
door = tlv.magnetic
while True:
    
    readcount += 1

    door_output = {
    "x" : -0.098,
    "y" : 0.392,
    "z" : 0.0,
    "reads" : readcount,
    }

    if (readcount % 100 == 1) :
        print('.', end='')
    if (readcount % 1000 == 1) :
        print("\n read: : ", readcount)
    mqtt_client.publish(DOOR_MQTT_TOPIC, json.dumps(door_output))

    time.sleep(.01)
```

**Comment by brentru** [comment]
What happens if you add a longer delay within the `while True` loop, something like 5 seconds or so?

**Comment by lubarb** [comment]
I haven't tried longer than 1 second. For what I'm trying to do, I need to be able to capture data about every .25 seconds.

I've tried with no delay at all, and observed that an error usually happens within a few seconds to a minute or so. With a .01 delay, it happens in a couple of minutes to several hours. With earlier code/libraries (so maybe now irrelevant), I had something that ran for 5-10 minutes up to a couple of days. So my unproven guess is that this occurs some small percentage of calls and that 5 seconds would give me hours to a week or so.

Also, I have two brokers I've tested with - a pi4 I just use for testing and an Intel sbc running home assistant, node red, and mqtt. The errors seem to occur much more quickly with the pi.

**Comment by lubarb** [comment]
I've started a 5 second test. At 90 reads and counting....

**Comment by brentru** [comment]
Looks like the connection is being closed. 

Could you add a `try/except` around your `publish` call:
```
try:
  mqtt_client.publish(DOOR_MQTT_TOPIC, json.dumps(door_output))
except OSError:
  mqtt_client.reconnect()
```

**Comment by lubarb** [comment]
It looks like you're right - the pi running the mqtt server has crashed. I'll rebuild that with another pi and also add the try/except you suggest. Thanks. I'll test & report back.

**Comment by lubarb** [comment]
The ssd from my original pi gave a verification error when I tried to reinstall Pi OS from scratch, followed quickly by refusing to respond to USB at all. Dead.

I loaded up a different pi & ssd with a fresh OS and mosquitto. I ran a 1 1/2 hr test with a .01 sec delay and the try/except recovered 4 times out of 250,000 publishes. I then ran overnight with a .25 sec delay with zero retries out of 170,000 publishes.

Errors on the pi ssd never occurred to me. Thanks for the help. Closing this issue.
